rust-runtime/aws-smithy-observability/src/meter.rs

```diff
@@ -0,0 +1,118 @@
+    fn get_meter(&self, scope: &'static str, attributes: Option<&Attributes>) -> Box<dyn Meter>;
+
+    /// Optional method to flush the metrics pipeline, default is noop
+    fn flush(&self) -> Result<(), ObservabilityError> {
+        Ok(())
+    }
+
+    /// Optional method to shutdown the metrics provider, default is noop
+    fn shutdown(&self) -> Result<(), ObservabilityError> {
+        Ok(())
+    }
+}
+
+/// The entry point to creating instruments. A grouping of related metrics.
+pub trait Meter {
+    /// Create a new Gauge.
+    #[allow(clippy::type_complexity)]
+    fn create_gauge(
+        &self,
+        name: String,
+        callback: Box<dyn Fn(&dyn AsyncMeasurement<Value = f64>) + Send + Sync>,
+        units: Option<String>,
+        description: Option<String>,
+    ) -> Box<dyn AsyncMeasurement<Value = f64>>;
+
+    /// Create a new [UpDownCounter].
+    fn create_up_down_counter(
+        &self,
+        name: String,
+        units: Option<String>,
+        description: Option<String>,
+    ) -> Box<dyn UpDownCounter>;
+
+    /// Create a new AsyncUpDownCounter.
+    #[allow(clippy::type_complexity)]
+    fn create_async_up_down_counter(
+        &self,
+        name: String,
+        callback: Box<dyn Fn(&dyn AsyncMeasurement<Value = i64>) + Send + Sync>,
+        units: Option<String>,
+        description: Option<String>,
+    ) -> Box<dyn AsyncMeasurement<Value = i64>>;
+
+    /// Create a new [MonotonicCounter].
+    fn create_monotonic_counter(
+        &self,
+        name: String,
+        units: Option<String>,
+        description: Option<String>,
+    ) -> Box<dyn MonotonicCounter>;
+
+    /// Create a new AsyncMonotonicCounter.
+    #[allow(clippy::type_complexity)]
+    fn create_async_monotonic_counter(
+        &self,
+        name: String,
+        callback: Box<dyn Fn(&dyn AsyncMeasurement<Value = u64>) + Send + Sync>,
+        units: Option<String>,
+        description: Option<String>,
+    ) -> Box<dyn AsyncMeasurement<Value = u64>>;
+
+    /// Create a new [Histogram].
+    fn create_histogram(
+        &self,
+        name: String,
+        units: Option<String>,
+        description: Option<String>,
+    ) -> Box<dyn Histogram>;
+}
+
+/// Collects a set of events with an event count and sum for all events.
+pub trait Histogram {
+    /// Record a value.
+    fn record(&self, value: f64, attributes: Option<&Attributes>, context: Option<&dyn Context>);
+}
+
+/// A counter that monotonically increases.
+pub trait MonotonicCounter {
+    /// Increment a counter by a fixed amount.
+    fn add(&self, value: u64, attributes: Option<&Attributes>, context: Option<&dyn Context>);
+}
+
+/// A counter that can increase or decrease.
+pub trait UpDownCounter {
+    /// Increment or decrement a counter by a fixed amount.
+    fn add(&self, value: i64, attributes: Option<&Attributes>, context: Option<&dyn Context>);
+}
+
+/// A measurement that can be taken asynchronously.
+pub trait AsyncMeasurement {
```

Comment:
I acknowledge there are many traits introduced in this PR, so I don't know how strictly we should enforce the trait naming convention in the codebase, i.e. verb + noun. For instance, `AsyncMeasurement` is probably on the "easy" side for renaming: `AsyncMeasure`. For others like `Histogram`, it's probably harder to rename. Maybe it's not straightforward to make names of new traits consistent, but this is just bringing awareness.

Reply:
I stuck with the names directly from the spec to align with other SDKs. I'm not strongly opposed to changing them, but we should discuss if breaking that alignment is worth it.

Reply:
For that, we've already broken alignments for the trait names in the SRA because the doc was mostly written with Java in mind (ex. the doc specifies `CredentialsProvider` as an interface but we have the `ProvideCredentials` trait to align with the verb + noun Rust ecosystem).

Given [John worked on a PR](https://github.com/smithy-lang/smithy-rs/pull/3065) just to rename trait names, personally I'd like to carry that spirit. If it's ok, let's make an effort to rename the obvious ones like `AsyncMeasure`, `MeterProvider` to verb + none to make them align with the codebase as much as possible. There are always exceptions like `HttpClient` trait in the existing code, and for them we can just leave them as they are (such as `Histogram`, `Meter`).

Reply:
Ack, I'll rearrange these to follow our grammatical convention in another commit since we have already committed to it with other SRA interfaces.